### PR TITLE
dev: migrate @woocommerce/components/link to ts

### LINKED ISSUE #35376: Migrate `@woocommerce/components/link` to TS
### Describe the solution you'd like

Migrate the Link component to TS

### Acceptance Criteria
- [ ] All `*.js` files are converted to `*.ts` or `*.tsx`
- [ ] Run `pnpm --filter=@woocommerce/components build` without any errors
- [ ] Run `pnpm --filter=woocommerce/client/admin ts:check` without any errors

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rjchow**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

- Renamed source, test and storybook files for the Link component to .tsx
- Added type annotations where necessary
- Added handling for undefined type errors in Link component, particularly around event.target.closest

Closes #35376.

### How to test the changes in this Pull Request:

1. Run `pnpm --filter=@woocommerce/components run test` - all unit tests should pass
1. Run `pnpm --filter=@woocommerce/storybook run storybook`
1. Go to http://localhost:6007/
1. Test `Link`
1. Smoke test around the components that Link is used in

image

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Review by moon0326 — APPROVED** [review]
Nice work on this @rjchow 👍 

LGTM and tested well 🚀

File: packages/js/components/src/link/index.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ * Use `Link` to create a link to another resource. It accepts a type to automatically
+ * create wp-admin links, wc-admin links, and external links.
+ */
+export const Link = ( {
+	href,
+	children,
+	type = 'wc-admin',
+	...props
+}: React.AnchorHTMLAttributes< HTMLAnchorElement > &
+	LinkProps ): React.ReactElement => {
```
Comment (chihsuan) [inline packages/js/components/src/link/index.tsx:33]: @rjchow What do you think if we use `React.FC`  instead?

Comment (rjchow) [reply packages/js/components/src/link/index.tsx:33]: We'd get type errors on the `<a {...passProps} />` when consumers try to put on props to passthrough that are valid for a but not specified on our LinkProps 😀

Comment (chihsuan) [reply packages/js/components/src/link/index.tsx:33]: Oh, I see.

**Review by chihsuan — APPROVED** [review]
Thanks for wrangling this! @rjchow  LGTM.

**Pull request merged**